src/dynamics/mpas/driver/dyn_mpas_subdriver.F90

```diff
@@ -0,0 +1,513 @@
+module dyn_mpas_subdriver
+    !-------------------------------------------------------------------------------
+    ! module dyn_mpas_subdriver
+    !
+    ! This module manages the life cycle (i.e., initialization, running, and
+    ! finalization) of MPAS as a dynamical core within CAM-SIMA.
+    !
+    !-------------------------------------------------------------------------------
+
+    use, intrinsic :: iso_fortran_env, only: output_unit
+
+    ! Modules from external libraries.
+    use mpi, only: mpi_comm_null, mpi_comm_rank, mpi_success
+    use pio, only: iosystem_desc_t, pio_iosystem_is_active
+
+    ! Modules from MPAS.
+    use atm_core_interface, only: atm_setup_core, atm_setup_domain
+    use mpas_derived_types, only: core_type, domain_type
+    use mpas_domain_routines, only: mpas_allocate_domain
+    use mpas_framework, only: mpas_framework_init_phase1, mpas_framework_init_phase2
+    use mpas_kind_types, only: strkind
+    use mpas_pool_routines, only: mpas_pool_get_config
+
+    implicit none
+
+    private
+    public :: mpas_dynamical_core_type, mpas_dynamical_core
+
+    abstract interface
+        ! This interface is compatible with `endrun` from CAM-SIMA.
+        subroutine model_error_if(message, file, line)
+            character(*),           intent(in) :: message
+            character(*), optional, intent(in) :: file
+            integer,      optional, intent(in) :: line
+        end subroutine model_error_if
+    end interface
+
+    !> The "class" of MPAS dynamical core.
+    !> Important data structures like states of MPAS dynamical core are encapsulated inside this derived type to prevent misuse.
+    !> Type-bound procedures provide well-defined APIs for CAM-SIMA to interact with MPAS dynamical core.
+    type :: mpas_dynamical_core_type
+        private
+
+        integer :: log_unit = output_unit
+        integer :: mpi_comm = mpi_comm_null
+        integer :: mpi_rank = 0
+        logical :: mpi_rank_master = .false.
+
+        ! Actual implementation is supplied at runtime.
+        procedure(model_error_if), nopass, pointer :: model_error => null()
+
+        type(core_type), pointer :: corelist => null()
+        type(domain_type), pointer :: domain_ptr => null()
+    contains
+        private
+
+        procedure, pass, public :: debug_print => dyn_mpas_debug_print
+        procedure, pass, public :: init_phase1 => dyn_mpas_init_phase1
+        procedure, pass, public :: read_namelist => dyn_mpas_read_namelist
+        procedure, pass, public :: init_phase2 => dyn_mpas_init_phase2
+    end type mpas_dynamical_core_type
+
+    !> The "instance/object" of MPAS dynamical core.
+    type(mpas_dynamical_core_type) :: mpas_dynamical_core
```

Comment:
I don't have a strong opinion (lacking some knowledge of the intended overall structure of a dycore in CAM-SIMA), but should this instance of `mpas_dynamical_core_type` be created elsewhere, perhaps in `dyn_comp.F90`? Supposing we might in some cases want multiple instances of the dycore, which module should "own" those corresponding instances of the `mpas_dynamical_core_type`?

Reply:
After some thoughts, yes, class instantiation and keeping it around should be user's responsibility.

I have moved instantiation of MPAS dynamical core to `dyn_comp` module. See commit https://github.com/ESCOMP/CAM-SIMA/pull/248/commits/c4fda967a804391bfd6c995b43ad89e02af99532.